MAVProxy/modules/mavproxy_DGPS.py

```diff
@@ -1,86 +1,208 @@
+        frame_size = 0
+        for id in sorted(self.id_counts.keys()):
+            print(" %4u: %u (len %u)" % (id, self.id_counts[id], len(self.last_by_id[id])))
+            frame_size += len(self.last_by_id[id])
+        print("dgps: %u packets, %.2f bytes/sec last %.3fs ago framesize %u" % (self.pkt_count, self.rate, now - self.last_pkt, frame_size))
+
+    def connect_udp_rtcm_base(self, ip, port):
+        print ("UDP Connection")
         self.portnum = 13320
         self.port = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
         self.port.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
         self.port.bind(("127.0.0.1", self.portnum))
         mavutil.set_close_on_exec(self.port.fileno())
         self.port.setblocking(0)
-        self.inject_seq_nr = 0
         print("DGPS: Listening for RTCM packets on UDP://%s:%s" % ("127.0.0.1", self.portnum))
-    
-    def send_rtcm_msg(self, data):
-        msglen = 180;
-        
-        if (len(data) > msglen * 4):
-            print("DGPS: Message too large", len(data))
-            return
-        
-        # How many messages will we send?
-        msgs = 0
-        if (len(data) % msglen == 0):
-            msgs = len(data) / msglen
-        else:
-            msgs = (len(data) / msglen) + 1
-        
-        for a in range(0, msgs):
-            
-            flags = 0
-            
-            # Set the fragment flag if we're sending more than 1 packet.
-            if (msgs) > 1:
-                flags = 1
-            
-            # Set the ID of this fragment
-            flags |= (a & 0x3) << 1
-            
-            # Set an overall sequence number
-            flags |= (self.inject_seq_nr & 0x1f) << 3
-            
-            
-            amount = min(len(data) - a * msglen, msglen)
-            datachunk = data[a*msglen : a*msglen + amount]
-            
-            self.master.mav.gps_rtcm_data_send(
-                flags,
-                len(datachunk),
-                bytearray(datachunk.ljust(180, '\0')))
-        
-        # Send a terminal 0-length message if we sent 2 or 3 exactly-full messages.     
-        if (msgs < 4) and (len(data) % msglen == 0) and (len(data) > msglen):
-            flags = 1 | (msgs & 0x3)  << 1 | (self.inject_seq_nr & 0x1f) << 3
-            self.master.mav.gps_rtcm_data_send(
-                flags,
-                0,
-                bytearray("".ljust(180, '\0')))
-            
-        self.inject_seq_nr += 1
-
 
+    def connect_tcp_rtcm_base(self, ip, port):
+        print ("TCP Connection")
+        try:
+            self.port = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            self.port.connect((ip, port))
+            self.port.settimeout(1)
+            self.port.setblocking(0)
+        except:
+            print ("ERROR: Failed to connect to RTCM base over TCP, retrying in 2.5s")
+            self.waiting = True
+            threading.Timer(2.5,self.connect_tcp_rtcm_base(ip, port)).start()
```

Comment:
I don't think we want a timer here, instead use the idle task to check for a pending connection. See the start_pending code in the mavproxy_ntrip module